recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,398 @@
+            )
+
+        return objectives["loss"].detach()
+
+    def evaluate_batch(self, batch, stage):
+        out = self.compute_forward(batch, stage=stage)
+        objectives = self.compute_objectives(out, batch, stage=stage)
+        return objectives["accuracy"].cpu()
+
+    def on_stage_end(self, stage, stage_loss, epoch=None):
+        if stage == sb.Stage.VALID:
+            logger.info(
+                f"Update: {self.optimizer_step} - valid_accuracy: {stage_loss:.3f}"
+            )
+            if self.hparams.use_wandb:
+                wandb.log(
+                    {
+                        "valid_acuracy": stage_loss,
+                    },
+                    step=self.optimizer_step,
+                )
+
+    def update_average(self, loss, avg_loss):
+        if avg_loss == 0.0:
+            avg_loss = loss.item()
+        else:
+            avg_loss = 0.99 * avg_loss + 0.01 * loss.item()
+        return avg_loss
+
+
+def sample_negatives(y, num_neg):
+    """
+       y is output of feature extractor
+    """
+    B, T, C = y.shape
+    high = T - 1
+    with torch.no_grad():
+        targets = torch.arange(T).unsqueeze(-1).expand(-1, num_neg).flatten()
+        neg_indcs = torch.randint(low=0, high=high, size=(B, T * num_neg))
+        # negative should not be target and to make distribution uniform shift all >
+        neg_indcs[neg_indcs >= targets] += 1
+
+    neg_indcs = neg_indcs + torch.arange(B).unsqueeze(1) * high
+    y = y.view(-1, C)
+    negs = y[neg_indcs.view(-1)]
+    negs = negs.view(B, T, num_neg, C).permute(2, 0, 1, 3)  # to N, B, T, C
+    return negs
+
+
+def dataio_prepare(hparams):
+    data_folder = hparams["data_folder"]
+
+    train_data = sb.dataio.dataset.DynamicItemDataset.from_csv(
+        csv_path=hparams["train_csv"], replacements={"data_root": data_folder},
+    )
+
+    # we sort training data to speed up training and get better results.
+    train_data = train_data.filtered_sorted(
+        sort_key="duration",
+        key_max_value={"duration": hparams["avoid_if_longer_than"]},
+        key_min_value={"duration": hparams["avoid_if_shorter_than"]},
+    )
+
+    valid_data = sb.dataio.dataset.DynamicItemDataset.from_csv(
+        csv_path=hparams["valid_csv"], replacements={"data_root": data_folder},
+    )
+    # We also sort the validation data so it is faster to validate
+    valid_data = valid_data.filtered_sorted(sort_key="duration")
+
+    datasets = [train_data, valid_data]
+
+    @sb.utils.data_pipeline.takes("wav")
+    @sb.utils.data_pipeline.provides("sig")
+    def audio_pipeline(wav):
+        info = torchaudio.info(wav)
+        sig = sb.dataio.dataio.read_audio(wav)
+        assert sig.dim() == 1, sig.dim()
+        resampled = torchaudio.transforms.Resample(
+            info.sample_rate, hparams["sample_rate"],
+        )(sig)
+        with torch.no_grad():
+            resampled = F.layer_norm(resampled, resampled.shape)
```

Comment:
Isn't this a bottleneck cpu-wise ?

Reply:
will change this is a mistake, only commonvoice needs this

Reply:
done